### DISCUSSION TITLE
POCL_EXTRA_BUILD_FLAGS env variable options

### DISCUSSION THREAD

**Discussion by HariAbram** [pocl/pocl#1484] [General]
When I try to set POCL_EXTRA_BUILD_FALGS, I get the following error. 

```
$ export POCL_EXTRA_BUILD_FLAGS="-mcpu=a64fx"
$ bin/main-intel-llvm -s 1024 --mat-mul 
running on ...
cpu-a64fx-0x001

terminate called after throwing an instance of 'sycl::_V1::compile_program_error'
  what():  The program was built for 1 devices
Build program log for 'cpu-a64fx-0x001':
Invalid build option: -mcpu=a64fx
 -30 (PI_ERROR_INVALID_VALUE)
Aborted (core dumped)
```

Since POCL builds the kernel with the LLVM clang, shouldn't the clang flag also work here? or am I missing something?

**Comment by franz** [comment]
Hello,

The `POCL_EXTRA_BUILD_FLAGS` only inserts extra flags in the `clBuildProgram` and `clCompileProgram` calls. These are still checked like all other options, so only valid OpenCL flags can be added with this env variable.

You can force the cpu model at build time by setting CMake variable `LLC_HOST_CPU` on the CMake command line, this will override autodetected CPU value.

**Reply by HariAbram** [reply]
Ah Okay. I already set LLC_HOST_CPU when I compiled POCL. I want POCL to vectorize the kernel across the work items, it is just unrolling loop rather than vectorizing, so is there an option of forcing the POCL code generator from the kernel to build it with vector instructions? I tried setting POCL_WORK_GROUP_METHOD to different values, but it is still not vectorizing, is there a way for me to set various flags during clBuildProgram() and see if that will vectorize it? such as -Ofast.

**Reply by franz** [reply]
PoCL relies on LLVM and its vectorizer passes to do the vectorizing. There is no option to force vectorization, but you can try some OpenCL options like `-cl-fast-relaxed-math` or `-cl-unsafe-math-optimizations` (Ofast is not an OpenCL option). These relax the floating-point restrictions and in some cases thats enough for the vectorizer.
